src/OrleansAzureUtils/MultiClusterNetwork/AzureTableBasedGossipChannel.cs

```diff
@@ -0,0 +1,202 @@
+        }
+
+        // IGossipChannel
+        public async Task Push(MultiClusterData data)
+        {
+            logger.Verbose("-Push data:{0}", data);
+
+            // this is typically called with just one, or very few, items in data
+            // thus, it is sufficiently performant to read rows individually
+
+            var retrievalTasks = new List<Task<GossipTableEntry>>();
+            if (data.Configuration != null)
+            {
+                retrievalTasks.Add(tableManager.ReadConfigurationEntryAsync());
+            }
+
+            foreach (var gateway in data.Gateways.Values)
+            {
+                retrievalTasks.Add(tableManager.ReadGatewayEntryAsync(gateway));
+            }
+
+            await Task.WhenAll(retrievalTasks);
+
+            var entriesFromStorage = retrievalTasks.Select(t => t.Result).Where(entry => entry != null);
+            await DiffAndWriteBack(data, entriesFromStorage); 
+        }
+
+        // IGossipChannel
+        public async Task<MultiClusterData> PushAndPull(MultiClusterData pushed)
+        {
+            logger.Verbose("-PushAndPull pushed:{0}", pushed);
+
+            try
+            {
+                var entriesFromStorage = await tableManager.FindAllGossipTableEntries();
+                var delta = await DiffAndWriteBack(pushed, entriesFromStorage);
+
+                logger.Verbose("-PushAndPull pulled delta:{0}", delta);
+
+                return delta;
+            }
+            catch (Exception e)
+            {
+                logger.Info("-PushAndPull encountered exception {0}", e);
+
+                throw e;
+            }
+        }
+
+        internal async Task<MultiClusterData> DiffAndWriteBack(MultiClusterData dataToPush, IEnumerable<GossipTableEntry> entriesFromStorage)
+        {
+            GossipTableEntry configRow = null;
+            var gateways = new Dictionary<SiloAddress, KeyValuePair<GatewayEntry, GossipTableEntry>>();
+            MultiClusterConfiguration returnedConfiguration = null;
+
+            // collect left-hand side data
+            MultiClusterConfiguration configToPush = dataToPush.Configuration;
+            foreach (var gatewayEntry in dataToPush.Gateways.Values)
+            {
+                if (!gatewayEntry.Expired)
+                    gateways[gatewayEntry.SiloAddress] = new KeyValuePair<GatewayEntry, GossipTableEntry>(gatewayEntry, null);
+            }
+
+            foreach (var tableEntry in entriesFromStorage)
+            {
+                if (tableEntry.RowKey.Equals(GossipTableEntry.CONFIGURATION_ROW))
+                {
+                    configRow = tableEntry;
+                    // interpret empty admin timestamp by taking the azure table timestamp instead
+                    // this allows an admin to inject a configuration by editing table more easily
+                    if (configRow.GossipTimestamp == default(DateTime))
```

Comment:
Yes, but it does not matter who wins, and e-tag conflicts are ignored, so I would not call it dangerous.